src/lib/audio/audio-util.js

```diff
@@ -59,9 +60,39 @@ const encodeAndAddSoundToVM = function (vm, samples, sampleRate, name, callback)
     });
 };
 
+const downsampleIfNeeded = (samples, sampleRate, resampler) => {
+    const duration = samples.length / sampleRate;
+    const encodedByteLength = samples.length * 2; /* bitDepth 16 bit */
+    // Resolve immediately if already within byte limit
+    if (encodedByteLength < SOUND_BYTE_LIMIT) {
+        return Promise.resolve({samples, sampleRate});
+    }
+    // If encodeable at 22khz, resample and call submitNewSamples again
+    if (duration * 22050 * 2 < SOUND_BYTE_LIMIT) {
+        return resampler({samples, sampleRate}, 22050);
```

Comment:
I do follow how the sampler is being passed to downsampleIfNeeded and then returned here, but that seems hard to follow. Is there any way to simplify it, given that there's only one resampler actually used in the code?

Reply:
The dependency injection here is mostly to make unit testing the logic of `downsampleIfNeeded` separate from the resampling process itself.

Reply:
that makes sense

Reply:
Here's another way you might handle it:

change the function declaration line from
```
const downsampleIfNeeded = (samples, sampleRate, resampler) => {
```
to
```
const downsampleIfNeeded = (samples, sampleRate, resampler = this.resampleBufferToRate) => {
```

...then, when you call `downsampleIfNeeded()` in your actual code, the context will be simpler for an outsider to understand, and you can still test `downsampleIfNeeded` the same way.

Do you like that idea?